### PR TITLE
Chore/create escrow

### PR DESCRIPTION
**Pull request by maxnorm**
- Add `create_escrow` instruction
- Add `CreateEscrowCtx` for the instruction context
- Add test for creating an escrow

I removed the accounts folder. We will add the instruction context with the instruction since it tightly related.

### CODE REVIEW

**Review by Captain-Cork — DISMISSED** [review]

File: anchor/tests/depo.spec.ts
[the remark this answers is not in the document]
Comment (maxnorm) [reply anchor/tests/depo.spec.ts:1]: I can. It won't be necessary for reference?

File: anchor/tests/depo.spec.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    )
+    .accounts({
+      escrow: escrowKey,
+      signer: initialiser.publicKey,
+      systemProgram: anchor.web3.SystemProgram.programId,
+    })
+    .signers([initialiser])
+    .rpc()
+
+    console.log('Created escrow with UUID:', uuid)
```
[the remark this answers is not in the document]
Comment (maxnorm) [reply anchor/tests/depo.spec.ts:67]: i added it! i forgot to do it

File: anchor/programs/depo/src/instructions/mod.rs [hunk 2]
```diff
@@ -1 +1,4 @@
-pub mod create_escrow;
\ No newline at end of file
+pub mod create_escrow;
```
[the remark this answers is not in the document]
Comment (maxnorm) [reply anchor/programs/depo/src/instructions/mod.rs:1]: I tried to do it but it's the only way i'm able to compile. 
image
image
image
image
image

File: anchor/tests/depo.spec.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+describe('DEPO', () => {
+  const provider = anchor.AnchorProvider.env()
+  anchor.setProvider(provider)
+  
+  const program = anchor.workspace.Depo as Program<Depo>
+
+  const initialiser = Keypair.generate()
+
+  beforeAll(async () => {
+    const sig = await provider.connection.requestAirdrop(
```
[the remark this answers is not in the document]
Comment (maxnorm) [reply anchor/tests/depo.spec.ts:18]: i'll make it more clear it for signature

File: anchor/tests/depo.spec.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      initialiser.publicKey,
+      10 * LAMPORTS_PER_SOL
+    )
+
+    await provider.connection.confirmTransaction(sig, 'confirmed')
+  
+    const balance = await provider.connection.getBalance(initialiser.publicKey)
+    console.log('Balance:', balance / LAMPORTS_PER_SOL, 'SOL')
```
[the remark this answers is not in the document]
Comment (maxnorm) [reply anchor/tests/depo.spec.ts:26]: it really depend since it the tests. It can be worth adding context with logs.

File: anchor/tests/depo.spec.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    if (balance === 0) throw new Error('Airdrop failed')
+  })
+
+  it('Create Escrow', async () => {
```
[the remark this answers is not in the document]
Comment (maxnorm) [reply anchor/tests/depo.spec.ts:30]: Good point! I thought of implementing the more instruction with 1 test and going back to add test case would be better for developing the program quicker? I can make them now too

File: anchor/programs/depo/src/instructions/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (guillaumedebavelaere) [reply anchor/programs/depo/src/instructions/mod.rs:1]: Ok let's stay like this, i will take a look later

File: anchor/tests/depo.spec.ts
[the remark this answers is not in the document]
Comment (guillaumedebavelaere) [reply anchor/tests/depo.spec.ts:1]: Not that now you added a test

**Review by guillaumedebavelaere — CHANGES_REQUESTED** [review]

File: anchor/tests/depo.spec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (guillaumedebavelaere) [reply anchor/tests/depo.spec.ts:26]: I mean I understand it can be useful for debugging, but then we should use assertions to test if we get what we expect ?

[the remark this answers is not in the document]
Comment (maxnorm) [reply anchor/tests/depo.spec.ts:26]: I'll remove them

File: anchor/tests/depo.spec.ts
[the remark this answers is not in the document]
Comment (guillaumedebavelaere) [reply anchor/tests/depo.spec.ts:1]: Yes 1 test file per instruction

[the remark this answers is not in the document]
Comment (maxnorm) [reply anchor/tests/depo.spec.ts:1]: good

**Review by guillaumedebavelaere — APPROVED** [review]
You can merge when the ticket for adding error test cases is created

File: anchor/programs/depo/src/instructions/create_escrow.rs [hunk 6]
```diff
@@ -1 +1,56 @@
...
+    escrow.description = description;
+
+    escrow.total_amount = 0;
+
+    escrow.is_public_deposit = true;
+    escrow.depositors_count = 0;
+    escrow.recipients_count = 0;
+
+    escrow.status = Status::Draft;
+    escrow.modules = Vec::new();
```
Comment (maxnorm) [inline anchor/programs/depo/src/instructions/create_escrow.rs:35]: I forgot to initialize the modules vec

File: anchor/programs/depo/src/instructions/create_escrow.rs [hunk 7]
```diff
@@ -1 +1,56 @@
...
+) -> Result<()> {
+    let escrow = &mut ctx.accounts.escrow;
+    escrow.id = escrow_id;
+    escrow.initialiser = ctx.accounts.signer.key();
+    escrow.name = name;
+    escrow.description = description;
+
+    escrow.total_amount = 0;
+
+    escrow.is_public_deposit = true;
```
Comment (maxnorm) [inline anchor/programs/depo/src/instructions/create_escrow.rs:30]: Forgot too

**Review by guillaumedebavelaere — APPROVED** [review]

File: anchor/tests/create_escrow_test.spec.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    expect(storedDescription).toBe(testDescription)
+    expect(escrowAccount.totalAmount.toNumber()).toBe(0)
+    expect(escrowAccount.isPublicDeposit).toBe(true)
+    expect(escrowAccount.depositorsCount).toBe(0)
+    expect(escrowAccount.recipientsCount).toBe(0)
+    expect(escrowAccount.status).toEqual({ draft: {} })
+    expect(escrowAccount.modules).toEqual([])
+  })
+
+  it.skip('Create Escrow with same UUID should fail', async () => {
```
Comment (maxnorm) [inline anchor/tests/create_escrow_test.spec.ts:82]: add edge case testing for later

File: anchor/tests/create_escrow_test.spec.ts [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+    )
+    .accounts({
+      escrow: escrowKey,
+      signer: initialiser.publicKey,
+      systemProgram: anchor.web3.SystemProgram.programId,
+    })
+    .signers([initialiser])
+    .rpc()
+    .catch((error) => {
+      expect(error).toBeDefined()
```
Comment (guillaumedebavelaere) [inline anchor/tests/create_escrow_test.spec.ts:130]: we should declare a variable and check this outside the catch otherwise if there is no error we won't see the test failing.

**Review by guillaumedebavelaere — APPROVED** [review]
